man/ucp.8
.\" $Header$
.\"
.\"  Man page for HUJI-NJE/FUNET-NJE on UNIX system.
.\"
.\"  Written by  Matti Aarnio <mea@finfiles.bitnet> <mea@nic.funet.fi>
.\"
.\"  Date: 27-Dec-1990, 18-Sep-1993, 11-Feb-1994
.\"
.fp 4 CW
.ds ]W Funet-Nje 8 UNIX 3.0
.TH "FUNET-NJE on UNIX" 8 "11-Feb 1994"
.SH NAME
ucp \- FUNETNJE bitnet communication program control function interface.
.SH SYNOPSIS
ucp
.RB [ " command to be issued " ]
.SH DESCRIPTION
.IX "ucp"
.IX "bitnet"
.IX "funetnje"
.LP
.I Ucp
command lets system manager to control
.BR funetnje (8)
functionality and modify some of parameters given on
configuration file (which usually is
.IR /etc/funetnje.cf .)
.LP
.I Ucp
can be started without any command line arguments, and then
it asks commands with
\f4NJE-CP>
prompt.
.LP
\f4ucp help\fR gives list of commands that
.I ucp
understands.
.SH SEE ALSO
.BR funetnje (8),
.BR mailify (1),
.BR send (1),
.BR sendfile (1),
.BR receive (1),
.BR bmail (8),
.SH DIAGNOSTICS
Obscure.  Most return results come with `mesg' facility, and user
.B must
have
.RB ` mesg (1) " y" '
to be able to receive those messages.
.SH NOTE
To be able to use this facility, you must have write permission to
the
.SM FUNET-NJE
inter-module communications `mailbox', which is implemented as
owner/group writable FIFO.
This program
.B should not
be turned into a SUID/SGID program, rather those who must be able
to command the funetnje shall be either super-user capable, or have
the access-group of the communications pipe.
See
.BR funetnje (8)
for more.
.SH AUTHORS
.LP
Yehavi Bourvine, Hebrew University of Yerusalem, Israel
.I <yehavi@hujivms.bitnet>.
.LP
Matti Aarnio, Finnish University and Research Network \- FUNET, Finland
.I <mea@nic.funet.fi> <mea@finfiles.bitnet>.
.LP
Original (VMS oriented package) software by Yehavi Bourvine,
UNIX features, plenty of modifications, man-pages by Matti Aarnio.


ChangeLog
Fri Mar 22 13:48:30 1996  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    Even more relaxed approach on parametrizing
	    the CMDMAILBOX -- now allows SOCK/FIFO/UDP
	    with fairly flexible (in regards of spaces)
	    syntax..

Mon Nov 13 17:48:52 1995  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    A bit more relaxed approach on parametrizing
	    CMDMAILBOX config parameter (FIFO/SOCKET case)

	* file_queue.c:
	    Do RSCS-accounting on locally sent files, that
	    wind up to local recipients

	* unix_brcst.c:
	    Re-enter the SysVR4 UTMP weirdos..  Now fixed ?

Fri Aug 25 09:02:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* mailify.c:
	    Handling error-messages (where "MAIL FROM:" address
	    is "<>", that is, empty) was done improperly.  Now
	    it does flag input as from "channel error" in that
	    case.

Mon Jul 24 14:29:59 1995  Matti Aarnio  <mea@nic.funet.fi>

	* bmail.c, sendfile.c:
	    Sometimes   mktemp()  fails when there are a LOT of
	    files being sent at the same time.  Changed now to
	    more classical  ".MaiL.%05d"-format with  getpid()
	    result as integer data..

Wed Jun 21 22:42:39 1995  Matti Aarnio  <mea@nic.funet.fi>

	* unix_brcst.c, nmr_unix.c:
	    Ok, more than __linux__ is needed to know which
	    machines have <utmp.h> containing extended flags.
	    Now we test  "DEAD_PROCESS" pre-processor define,
	    which existence (propably) will correlate with
	    the existence of utmp's field  ut_type.
	    (SysVr4 feature, I think.)

Mon Jun 19 15:17:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* send.c, sendfile.c, bmail.c:
	    Some additional tests to trap various forms of bad
	    addresses -- accepts only  user@node format with
	    no '.' in the address, nor either component longer
	    than 8 chars.

	* nmr.c:
	    Made sure not to overwrite allocated string spaces on
	    Msg-queue entry.  (Did cause a crash with it..)

Tue Jun 13 09:13:18 1995  Matti Aarnio  <mea@nic.funet.fi>

	* qrdr.c:
	    Collect directory browsing results to an array at first,
	    then qsort() it with the time-stamp, and finally display
	    them in order. (Latest last)
	    (This is something what Gerald Hanush asked for.)

Mon Jun 12 22:44:00 1995  Matti Aarnio  <mea@nic.funet.fi>

	* file_queue.c:
	    Apparently OSF/1 pipe() return value is the first
	    fd, not "0", when it succeeds.  Nice thing - and not..
	    Altered the test to see if it returns a value under
	    zero, or something positive.  When the return is
	    positive (zero, or larger), fork() and run in
	    child-mode.

	    If a child_pipe exists already ( >= 0 ), close
	    it and mark the new one for processing.

	* libreadcfg.c, read_config, libsendcmd.c, unix.c:
	    Rewrote things so that now the used CMDMAILBOX
	    transport method is chosen at runtime:

	    CMDMAILBOX  F  /path/to/.cmdfifo
	    CMDMAILBOX  f  /path/to/.cmdfifo
	    CMDMAILBOX  S  /path/to/.cmdsocket
	    CMDMAILBOX  U  127.0.0.2 175	# UDP port

	    The "f" is FIFO which is closed, and reopened
	    every time one command is pulled in.

Sun Jun 11 03:12:06 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesparser.c:
	    Noticed that the data parsing stopped at first ":",
	    which caused premature breaks; observe following entry
	    from BITEARN.NODES -file:
	    	:remark1.BASE(9501) 00/00/00 00:00:00

	* libreadcfg.c, read_config.c, libsendcmd.c, unix.c:
	   Amended "CMDMAILBOX"-data to contain port-number in
	   case UDP-format is used. (That is, have two tokens.)
	   Default value for the port-number is 175.

	* unix_tcp.c read_passive_tcp_connection():
	   Noticed that a case when a socket exists, and a new
	   connection is attempted is handled improperly. The
	   old socket-fd was FORGOTTEN!  Now it NAKs the connection
	   attempt with a complaint that there already is a connection
	   in there...

	   Things will progress from such a situation usually by
	   detecting a closure on the existing stream, but better
	   not to rush ahead of things.

	* bmail.c:
	   Accept "-u" only when calling the program from uids 0 or 1
	   (root and daemon in most common cases -- for BSD-sendmail
	    which wants  bmail executable with protection 2111 (user X,
	    group NJEGRP) )

Sat Jun 10 05:18:27 1995  Matti Aarnio  <mea@nic.funet.fi>

	* Makefile, consts.h, prototypes.h, recv_file.c, io.c,
	  send_file.c, unix.c, protocol.c, unix_tcp.c:

	    - Found one uninitialized pointer, which apparently
	      caused mystic data-structure corruptions (on timer
	      subsystem)

	    - Made mandatory (well, with disabling options) to
	      use NBSTREAM, USE_XMIT_QUEUE, and USE_SOCKOPT.
	      (in consts.h, see also  Makefile's  comments!)

	    - Use  F_WAIT_V_A_BIT  to flag if there is any reason
	      for NOT to send out data at the moment, and instead
	      quit from within  handle_ack()..
	      The F_WAIT_V_A_BIT is set when new data is sent to
	      outbound buffer (primary one, or XMIT QUEUE).
	      The bit is cleared when pending writer finishes
	      with success status.  When the pending-writer is
	      called from  poll_sockets(), immediately after it
	      the  dequeue_xmit_queue() is called, and if there
	      is something, it can cause F_WAIT_V_A_BIT to be
	      set again..)

	    - If the XMIT_QUEUE ever runs out of space, it abort()s!

	    - Tested the beast by sending 80 or so smallish files both
	      ways over the link.  Such setup caused previously link
	      corruption..  Actually the critical phase is when there
	      are flowing NJE headers/trailer, bare data sending works
	      just fine, but when a file ends, EOF block needs an ack
	      to be sent, and unless there the care is used, all
	      queues will overflow..

Thu Jun  8 15:48:14 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesfilter.c:
	    New option: "-netsoft" to list node software fields.

	* funetnje.texinfo:
	    Some additional chapters

	* unix_tcp.c:

ucp.c
/* UCP.C (formerly CP.C)     V1.8-mea
 | Copyright (c) 1988,1989,1990 by
 | The Hebrew University of Jerusalem, Computation Center.
 |
 |   This software is distributed under a license from the Hebrew University
 | of Jerusalem. It may be copied only under the terms listed in the license
 | agreement. This copyright message should never be changed or removed.
 |   This software is gievn without any warranty, and the Hebrew University
 | of Jerusalem assumes no responsibility for any damage that might be caused
 | by use or misuse of this software.
 |
 | Control program for the RSCS daemon. It assigns a channel to its mailbox
 | and send a command message to it.

 | Altered outlook, more features, and options..
 | Copyright (c) 1991, 1993, 1994,  Finnish University and Research Network,
 | FUNET.
 
 */

#include "prototypes.h"
#include "clientutils.h"

#define LINESIZE 256

#define	Exit	1
#define	Show	2
#define	Shut	3
#define	Start	4
#define	Help	5
#define	Queue	6
#define	Stop	7
#define	Force	8
#define	Debug	9
#define	Loglevel 10
#define	Route	11
#define	Gone	12
#define	Ungone	13
#define Rescan	14

struct Commands {
	char	*command;
	int	code;
	} commands[] = {
	  { "EXIT", Exit },
	  { "SHOW", Show },
	  { "SHUT", Shut },
	  { "START", Start },
	  { "HELP", Help },
	  { "QUEUE", Queue },
	  { "STOP", Stop },
	  { "FORCE", Force },
	  { "DEBUG", Debug },
	  { "LOGLEVEL", Loglevel },
	  { "ROUTE", Route },
	  { "GONE", Gone },
	  { "UNGONE", Ungone },
	  { "RESCAN", Rescan },
	  { "", 0 }
	};

extern void process_cmd();
extern void show_help();
extern int parse_command();
extern void command();


char LOCAL_NAME   [10];
char BITNET_QUEUE [80];
char LOG_FILE     [80] = "-";

int LogLevel = 1;
FILE *LogFd = NULL;

int
main(cc, vv)
int	cc;
char	**vv;
{
	char	line[LINESIZE];
	int	i;

	read_configuration();

	if (cc == 1) {
	  while (!feof(stdin) && !ferror(stdin)) {
	    printf("NJE-CP> ");
	    if (fgets(line, sizeof line, stdin) == NULL)
	      exit(0);
	    process_cmd(line);
	  }
	} else {
	  *line = '\0';
	  for (i = 1; i < cc; i++) {
	    strcat(line, vv[i]);
	    strcat(line, " ");
	  }
	  process_cmd(line);
	}
	exit(0);
}

void
process_cmd(line)
char	*line;
{
	char	*p, KeyWord[80], param1[80], param2[80], param3[80];
	char	TempStr[160];
	int	i, NumParams, cmd;

	if ((p = strchr(line, '\n')) != NULL) *p = '\0';
	if (*line == '\0') return;	/* Null command */
	*param1 = 0;
	*param2 = 0;
	*param3 = 0;
	NumParams = sscanf(line, "%s %s %s %s", KeyWord, param1,
		param2, param3);
	switch ((cmd = parse_command(KeyWord))) {
	  case Help:
	      show_help();
	      break;
	  case Exit:
	      exit(0);	/* Exit program */
	  case Show:
	      if (NumParams != 2) {
		printf("Illegal SHOW command\n");
		break;
	      }
	      if (strncasecmp(param1, "LINE", 4) == 0)
		command(CMD_SHOW_LINES);
	      else
		if (strncasecmp(param1, "QUEUE", 5) == 0)
		  command(CMD_SHOW_QUEUE);
		else
		  printf("Illegal SHOW command\n");
	      break;
	  case Start:
	      if (NumParams != 3) {
		printf("Illegal START command\n");
		break;
	      }
	      sscanf(param2, "%d", &i);
	      if (strncasecmp(param1, "LINE", 4) == 0)
		command(CMD_START_LINE, i);
	      else
		printf("Illegal START command\n");
	      break;
	  case Force:
	  case Stop:
	      if (NumParams != 3) {
		printf("Illegal STOP/FORCE command\n");
		break;
	      }
	      if (strncasecmp(param1, "LINE", 4) != 0) {
		printf("Illegal STOP command\n");
		break;
	      }
	      if (sscanf(param2, "%d", &i) != 1) {
		printf("Illegal line number\n");
		break;
	      }
	      if (cmd == Force)
		command(CMD_FORCE_LINE, i);
	      else
		command(CMD_STOP_LINE, i);
	      break;
	  case Gone:
	  case Ungone:
	      if (cmd == Gone) {
		if (NumParams != 3) {
		  printf("Illegal GONE command\n");
		  break;
		}
		sprintf(line, "%s %s", param1, param2);
		command(CMD_GONE_ADD_UCP, 0, line);
	      } else {
		if (NumParams != 2) {
		  printf("Illegal GONE command\n");
		  break;
		}
		command(CMD_GONE_DEL_UCP, 0, param1);
	      }
	      break;
	  case Queue:
	      if (NumParams < 2) {
		printf("No filename given\n");
		break;
	      }
	      if ((NumParams == 4) &&
		 (strncasecmp(param2, "SIZE", 4) == 0))
		sscanf(param3, "%d", &i);
	      else
		i = 0;
	      command(CMD_QUEUE_FILE, 0, param1, i);
	      break;
	  case Shut:
	      if (NumParams == 1)	/* Shutdown normally */
		command(CMD_SHUTDOWN);
	      else
		if (strncasecmp(param1, "ABORT", 5) == 0)
		  command(CMD_SHUTDOWN_ABRT);
		else
		  printf("Illegal SHUT command\n");
	      break;
	  case Debug:
	      if (NumParams != 2 && NumParams != 3) {
		printf("DEBUG must get a parameter\n");
		break;
	      }
	      if (strncasecmp(param1, "DUMP", 4) == 0)
		command(CMD_DEBUG_DUMP);
	      else
		if (strncasecmp(param1, "RESCAN", 6) == 0)
		  command(CMD_DEBUG_RESCAN,0,param2,0);
		else
		  printf("Illegal DEBUG command\n");
	      break;
	  case Rescan:
	      if (NumParams < 2) {
		printf("  RESCAN QUEUE, or RESCAN EXITS ?\n");
		break;
	      }
	      if (strcasecmp(param1,"QUEUE")==0) {
		if (*param2 == 0) *param2 = '+';
		command(CMD_DEBUG_RESCAN,0,param2,0);
	      } else if (strcasecmp(param1,"EXITS")==0) {
		command(CMD_EXIT_RESCAN,0,0,0);
	      } else if (strcasecmp(param1,"ROUTE")==0) {
		command(CMD_ROUTE_RESCAN,0,0,0);
	      } else {
		printf("  RESCAN EXITS, RESCAN QUEUE, or RESCAN ROUTE ?\n");
	      }
	      break;
	  case Loglevel:
	      if (NumParams != 2) {
		printf("No log level given\n");
		break;
	      }
	      sscanf(param1, "%d", &i);
	      command(CMD_LOGLEVEL, i);
	      break;
	  case Route:
	      if (NumParams != 4) {
		printf("Illegal format: Try: ROUTE xx TO yy\n");
		break;
	      }
	      if (strncasecmp(param2, "TO", 2) != 0) {
		printf("Illegal format: Try: ROUTE xx TO yy\n");
		break;
	      }
	      sprintf(TempStr, "%s %s", param1, param3);
	      command(CMD_CHANGE_ROUTE, 0, TempStr, 0);
	      break;
	  default:
	      printf("Illegal command '%s'. Type HELP\n", line);
	      break;
	}
}

void
show_help()
{
	printf("   HELP  - Show this message\n");
	printf("   SHOW LINE/QUEUE - Show lines or queue status\n");
	printf("   START LINE n - Start a closed line\n");
	printf(" * START STREAM n LINE m - Start specific stream in active line\n");
	printf("   SHUT [ABORT] - Shutdown or abort the whole program\n");
	printf("   STOP LINE - Stop a line\n");
	printf("   FORCE LINE - Stop a line immediately\n");
	printf(" * STOP STREAM n LINE m - Stop a single stream in active line\n");
	printf(" * FORCE STREAM n LINE m - Stop immediately\n");
	printf("   QUEUE file-name [SIZE size] - To queue a file to send\n");
	printf("   RESCAN EXITS - Rescan file and message exits.\n");
	printf("   RESCAN QUEUE - Rescan queue and requeue files.\n");
	printf("   RESCAN ROUTE - Reopen route database.\n");
	printf("   DEBUG DUMP - Dump all lines buffers to logfile\n");
	printf(" %% DEBUG RESCAN - Rescan queue and requeue files.\n");
	printf("   LOGLEVEL n - Set the loglevel to N\n");
	printf("   ROUTE xxx TO yyy - Change the routing table.\n");
	printf("        Route to  OFF  will delete the route entry.\n");
	printf("   GONE username LoginDirectory - Add username to gone list\n");
	printf("   UNGONE username - Remove a user from the Gone list.\n");
	printf("* - Not yet implemented,  %% - obsoleted\n");
}


/*
 | Send the command. The command is in CMD, a numeric value (if needed) is in
 | VALUE, and a string value (if needed) is in STRING.
 */
void
command(cmd, value, string, param)
int	cmd, value, param;
char	*string;
{
#ifdef VMS
	struct	DESC	MailBoxDesc;
#endif
	char	line[LINESIZE];
	long	size;

	*line = cmd;
	switch (cmd) {
	case CMD_QUEUE_FILE :
			/* File size: */
		line[1] = (unsigned char)((param & 0xff00) >> 8);
		line[2] = (unsigned char)(param & 0xff);
		strcpy(&line[3], string);
		size = strlen(&line[3]) + 3;
		break;
	case CMD_CHANGE_ROUTE:
		strcpy(&line[1], string);
		size = strlen(&line[1]) + 1;
		break;
	case CMD_START_LINE:
	case CMD_STOP_LINE:
	case CMD_FORCE_LINE:
	case CMD_LOGLEVEL:
		line[1] = (value & 0xff); line[2] = '\0';
		size = 2;
		break;
	case CMD_GONE_ADD_UCP:
	case CMD_GONE_DEL_UCP:
		strcpy(&line[1], string);
		size = strlen(string) + 1;
		break;
	case CMD_DEBUG_RESCAN:
		mcuserid(&line[1]);	/* Add the username of sender */
		line[11] = *string;
		size = 13;
		break;
	default:
		mcuserid(&line[1]);	/* Add the username of sender */
		size = strlen(&line[1]) + 1;	/* 1 for the command code */
		break;
	}

	send_cmd_msg(line, size, 0); /* Must be ONLINE! */
}



int
parse_command(line)
char	*line;
{
	int	i;

	for (i = 0; ; i++) {		/* Look for the command */
	  if (commands[i].code == 0) /* End of valid commands */
	    return 0;		/* Signal error */
	  if (strncasecmp(commands[i].command, line,
			  strlen(line)) == 0)
	    return commands[i].code;
	}
}
